snapcraft/pluginhandler.py

```diff
@@ -301,7 +359,21 @@ def _new_clean(self, project_staged_state=None,
         if not project_stripped_state:
             project_stripped_state = {}
 
-        self._clean_steps(project_staged_state, project_stripped_state, step)
+        try:
+            self._clean_steps(project_staged_state, project_stripped_state,
+                              step)
+        except MissingState:
+            # If one of the step cleaning rules is missing state, it must be
+            # running on the output of an old Snapcraft. In that case, if we
+            # were specifically asked to clean that step we need to fail.
+            # Otherwise, just clean like the old Snapcraft did, and blow away
+            # the entire part directory.
+            if step:
+                raise
+
```

Comment:
Well, running on old Snapcraft output will probably be a common case. I can't think of anything to print that wouldn't make users suspicious that something is wrong when it's not.